pkg/controller/sidecarset/sidecarset_utils.go

```diff
@@ -101,3 +112,182 @@ func inconsistentStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha
+	// value is sidecarset hash
+	podSidecarUpdated map[string]string
+}
+
+func (u *updatedPodCache) set(key, hash string) {
+	u.lock.Lock()
+	u.podSidecarUpdated[key] = hash
+	u.lock.Unlock()
+}
+
+func (u *updatedPodCache) reset() {
+	u.lock.Lock()
+	u.podSidecarUpdated = make(map[string]string)
+	u.lock.Unlock()
+}
+
+// return true means: sidecar of pod is updated
+func (u *updatedPodCache) isUpdated(key, curHash string) bool {
+	u.lock.RLock()
+	hash := u.podSidecarUpdated[key]
+	u.lock.RUnlock()
+
+	if hash == curHash {
+		return true
+	}
+	if hash != "" && hash != curHash {
+		u.reset()
+		return false
+	}
+	return false
+}
+
+// available definition:
+// 1. image in pod.spec and pod.status is exactly the same
+// 2. pod is ready
+func getUnavailableNumber(sidecarSet *appsv1alpha1.SidecarSet, pods []*corev1.Pod) (int, error) {
+	var unavailableNum int
+	for _, pod := range pods {
+		// in case of informer cache latency
+		podInCacheUpdated := updateCache.isUpdated(
+			fmt.Sprintf("%v/%v/%v", sidecarSet.Name, pod.Namespace, pod.Name),
+			sidecarSet.Annotations[sidecarsetmutating.SidecarSetHashAnnotation])
+		podInInformerUpdated, err := isPodUpdated(sidecarSet, pod)
+		if err != nil {
+			return 0, err
+		}
+		if podInCacheUpdated && !podInInformerUpdated {
+			unavailableNum++
+			continue
+		}
+
+		if !isPodImageConsistent(pod) {
+			unavailableNum++
+			continue
+		}
+
+		if !isRunningAndReady(pod) {
+			unavailableNum++
+		}
+	}
+	return unavailableNum, nil
+}
+
+func isPodImageConsistent(pod *corev1.Pod) bool {
+	containerSpecImage := make(map[string]string)
+	for _, container := range pod.Spec.Containers {
+		containerSpecImage[container.Name] = container.Image
+	}
+
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if containerSpecImage[containerStatus.Name] != containerStatus.Image {
+			return false
+		}
+	}
+	return true
+}
+
+func onlyImageChanged(sidecarSet *appsv1alpha1.SidecarSet, pods []*corev1.Pod) ([]*corev1.Pod, error) {
+	var filteredPods []*corev1.Pod
+	hashKey1 := sidecarsetmutating.SidecarSetHashAnnotation
+	hashKey2 := sidecarsetmutating.SidecarSetHashWithoutImageAnnotation
+	for _, pod := range pods {
+		sidecarSetChanged, err := hashChanged(pod.Annotations[hashKey1], sidecarSet.Name, sidecarSet.Annotations[hashKey1])
+		if err != nil {
+			return nil, err
+		}
+		sidecarSetWithoutImageChanged, err := hashChanged(pod.Annotations[hashKey2], sidecarSet.Name, sidecarSet.Annotations[hashKey2])
+		if err != nil {
+			return nil, err
+		}
+
+		if sidecarSetChanged && !sidecarSetWithoutImageChanged {
+			filteredPods = append(filteredPods, pod)
+		}
+	}
+	return filteredPods, nil
+}
+
+func hashChanged(podHashs, targetName, targetHash string) (bool, error) {
+	sidecarSetHash := make(map[string]string)
+	if err := json.Unmarshal([]byte(podHashs), &sidecarSetHash); err != nil {
+		return false, err
+	}
+	return sidecarSetHash[targetName] != targetHash, nil
```

Comment:
It is possible that the Pod does not have any annotation. (e.g., A matching Pod created before SetcarSet is created) . So sidecarSetHash can be an empty map and the hashChanged method can return true. We should exclude this kind of Pods from filtered Pods.

Reply:
done, add isPodCreatedBeforeSidecarSet function